### PR TITLE
Add TypeId and CompactId caches to Communicator

### PR DESCRIPTION
**Pull request by externl**
Also remove per InputStream caches

### CODE REVIEW

File: csharp/src/Ice/Communicator.cs [hunk 1]
```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
...
-                foreach (string ns in _typeIdNamespaces)
+                // First attempt corresponds to no cs:namespace metadata in the
+                // enclosing top-level module
+                string className = TypeToClass(id);
+                Type? classType = AssemblyUtil.FindType(className);
+
+                //
+                // If this fails, look for helper classes in the typeIdNamespaces namespace(s)
+                //
+                if (classType == null && _typeIdNamespaces != null)
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs]: I think we can remove the `_typeIdNamespaces != null` check `_typeIdNamespaces` is non null reference

File: csharp/src/Ice/Communicator.cs [hunk 2]
```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
...
-                    if (helper != null)
+                    foreach (string ns in _typeIdNamespaces)
                     {
-                        try
-                        {
-                            c = helper.GetProperty("targetClass").PropertyType;
-                            break; // foreach
-                        }
-                        catch (System.Exception)
+                        Type? helper = AssemblyUtil.FindType(ns + "." + className);
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs]:
```suggestion
                        Type? helper = AssemblyUtil.FindType($"{ns}.{className}");
```

File: csharp/src/Ice/Communicator.cs [hunk 3]
```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
...
+                        if (helper != null)
                         {
+                            try
+                            {
+                                classType = helper.GetProperty("targetClass").PropertyType;
+                                break; // foreach
+                            }
+                            catch (System.Exception)
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs]:
```suggestion
                            catch (Exception)
```

File: csharp/src/Ice/Communicator.cs [hunk 4]
```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
...
-            if (_typeIdNamespaces != null)
+            return _compactIdCache.GetOrAdd(compactId, id =>
             {
-                compactIdNamespaces.AddRange(_typeIdNamespaces);
-            }
+                string[] defaultVal = { "IceCompactId" };
+                var compactIdNamespaces = new List<string>(defaultVal);
 
-            string? result = null;
+                if (_typeIdNamespaces != null)
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs]: no need for the null check here

File: csharp/src/Ice/Communicator.cs [hunk 5]
```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
...
+                        {
+                            var result = (string)classType.GetField("typeId").GetValue(null);
+                            if (result != null)
+                            {
+                                return ResolveClass(result);
+                            }
+                            return null;
+                        }
+                    }
+                    catch (System.Exception)
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs]:
```suggestion
                    catch (Exception)
```

File: csharp/src/Ice/Communicator.cs [hunk 6]
```diff
@@ -179,6 +180,8 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
         private int _state;
         private readonly IceInternal.Timer _timer;
         private readonly string[] _typeIdNamespaces = { "Ice.TypeId" };
```
Comment (bentoi) [inline csharp/src/Ice/Communicator.cs]: Do we need to initialize this here given that we also initialize it in the constructor?

File: csharp/src/Ice/Communicator.cs [hunk 7]
```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
...
-            string[] defaultVal = { "IceCompactId" };
-            var compactIdNamespaces = new List<string>(defaultVal);
-
-            if (_typeIdNamespaces != null)
+            return _compactIdCache.GetOrAdd(compactId, id =>
             {
-                compactIdNamespaces.AddRange(_typeIdNamespaces);
-            }
+                string[] defaultVal = { "IceCompactId" };
+                var compactIdNamespaces = new List<string>(defaultVal);
```
Comment (bentoi) [inline csharp/src/Ice/Communicator.cs]: Should we cache compacIdNamespaces like we cache typeIdNamespaces?

File: csharp/src/Ice/Communicator.cs (hunk 6 above)
Comment (externl) [reply csharp/src/Ice/Communicator.cs]: No, I'll remove it.

**Comment by externl** [comment]
> What are we doing with the compact ID resolver? We still accept one on communicator construction but no longer use it.

We don't use the resolver, I'll remove it.

File: csharp/src/Ice/Communicator.cs (hunk 7 above)
Comment (externl) [reply csharp/src/Ice/Communicator.cs]: I think we should!

**Pull request merged**